keps/sig-instrumentation/2580-custom-metrics-router/README.md

```diff
@@ -0,0 +1,651 @@
+create a router which acts as the single entry point for all metrics which
+then routes the metric request to the correct metrics provider based on
+existing information already exposed in most metrics adapters. The router can
+leverage existing projects like the
+[custom-metrics-apiserver](https://github.com/kubernetes-sigs/custom-metrics-apiserver)
+and the [metrics client](https://github.com/kubernetes/metrics).
+
+### Goals
+
+* Allows for multiple metric sources in a cluster Existing metrics sources
+* should be compatible with solution Leverage existing functionality in metric
+* sources for discovery and routing.
+
+### Non-Goals
+
+* Configuring metric sources at a fine-grain level
+
+## Proposal
+
+### Background
+
+#### custom-metrics-apiserver
+This [project](https://github.com/kubernetes-sigs/custom-metrics-apiserver)
+provides helper classes to build custom/external metric providers. Most
+custom/external metric providers use this framework. It is important to note
+here that the metrics server is an aggregated API server which provides a lot
+of the functionality which is present in the kube-apiserver like API
+discovery and RBAC authentication/authorization.
+
+#### Metrics Client
+The [API and client](https://github.com/kubernetes/metrics) for resource,
+custom and external metrics are used by the HPA controller to query for
+metrics from any registered custom/external metric providers. Technically the
+client is instantiated with the kube-apiserver as the target for the requests
+which in turn forwards it to the metrics server based on the API group i.e.
+`metrics.k8s.io`, `custom.metrics.k8s.io` or `external.metrics.k8s.io`.
+
+#### API Discovery
+API servers expose endpoints which can be used by a client to find the
+correct resource version and discover information about which API resources
+are provided by the aggregated API server. The _custom-metrics-apiserver_
+similarly exposes an endpoint which lists all the custom metric types(the
+associated resource type and metric name) and external metrics types which
+are provided by the metrics server.
+
+### Solution
+The proposed solution consists of 3 main components. A controller which uses
+a CRD to track metric sources. The controller uses API discovery to find the
+metrics provided by different metrics providers. It then builds a routing
+table mapping the metric to a provider. The second component is the routing
+table. One of the requirements of this routing table is that routing entries
+should be able to be added and deleted without a complete re-indexing of all
+metrics. And finally there is the request router. This component receives the
+metric requests which are proxied from the API server and then consults the
+routing table and queries the correct metric source and returns the result.
+
+![Router Architecture](custom_metrics_router.png)
+
+#### Metrics Discovery Controller
+`CustomMetricsSource` is the proposed CRD which should be used to register
```

Comment:
The current name could give users an impression, that it servers custom metrics only. What about `MetricsSourceProvider` or a similar name?

Reply:
I'm open to other names. I wanted to keep the name similar to the `APIService` object already present and with the `custom-metrics-apiserver`.

Reply:
Yeah, I just would like to avoid the `Custom` prefix, since this should deal with external metrics as well, otherwise I don't have a strong opinion on the name. 
`APIService` objects are named `v1beta[12].custom.metrics.k8s.io` or `v1beta1.external.metrics.k8s.io`, that depends on the type of metrics.